crates/runtime/src/http/v1/query.rs

```diff
@@ -102,18 +105,379 @@ use super::{sql_to_http_response, ArrowFormat};
+        )))
+        )
+    )
+)]
+pub(crate) async fn prepared(
+    Extension(df): Extension<Arc<DataFusion>>,
+    accept: Option<TypedHeader<Accept>>,
+    Json(PreparedStatement { sql, params }): Json<PreparedStatement>,
+) -> Response {
+    let params = match convert_json_to_param_values(params) {
+        Ok(params) => params,
+        Err(e) => {
+            tracing::debug!("Error reading params: {e}");
+            return (StatusCode::BAD_REQUEST, e.to_string()).into_response();
+        }
+    };
+
+    sql_to_http_response(
+        df,
+        &sql,
+        Some(params),
+        ArrowFormat::from_accept_header(accept.as_ref()),
+    )
+    .await
+}
+
+/// Converts a serde_json::Value into a datafusion::common::ParamValues.
+///
+/// # Arguments
+/// * `json` - The JSON value to convert.
+///
+/// # Returns
+/// * `Result<ParamValues, String>` - The converted ParamValues or an error message.
+///
+/// # Supported JSON Formats
+/// - Array: Converted to positional parameters (ParamValues::Positional).
+/// - Object: Converted to named parameters (ParamValues::Named).
+/// - Primitive types (null, bool, number, string) are supported within arrays or objects.
+///
+/// # Errors
+/// - Returns an error if the top-level JSON is not an array or object.
+/// - Returns an error if any value cannot be converted to a ScalarValue.
+fn convert_json_to_param_values(json: Value) -> Result<ParamValues, String> {
+    match json {
+        Value::Array(arr) => {
+            let mut vec = Vec::with_capacity(arr.len());
+            for (i, val) in arr.into_iter().enumerate() {
+                let scalar = json_to_scalar(&val).map_err(|e| {
+                    format!("failed to convert array element at index {}: {}", i, e)
+                })?;
+                vec.push(scalar);
+            }
+            Ok(ParamValues::List(vec))
+        }
+        Value::Object(obj) => {
+            let mut map = HashMap::new();
+            for (key, val) in obj {
+                let scalar = json_to_scalar(&val)
+                    .map_err(|e| format!("failed to convert value for key '{}': {}", key, e))?;
+                map.insert(key, scalar);
+            }
+            Ok(ParamValues::Map(map))
+        }
+        _ => Err("params must be a JSON array or object".to_string()),
+    }
+}
+
+/// Helper function to convert a single serde_json::Value to a ScalarValue.
+///
+/// # Arguments
+/// * `json` - The JSON value to convert.
+///
+/// # Returns
+/// * `Result<ScalarValue, String>` - The converted ScalarValue or an error message.
+///
+/// # Supported JSON Types
+/// - Null -> ScalarValue::Utf8(None)
+/// - Bool -> ScalarValue::Boolean(Some(bool))
+/// - Number -> ScalarValue::Int64(Some(i64)) if integer, else ScalarValue::Float64(Some(f64))
+/// - String -> ScalarValue::Utf8(Some(String))
+/// - Arrays and objects return an error (handled at a higher level).
+fn json_to_scalar(json: &Value) -> Result<ScalarValue, String> {
```

Comment:
This is probably useful outside of the HTTP module context. Let's find a more general place for this to live. I would also double check that this functionality doesn't already exist in the DataFusion crate somewhere - it seems like a common thing to want to do?

Reply:
This should also use Snafu errors